boa/src/syntax/parser/cursor.rs

```diff
@@ -0,0 +1,110 @@
+//! Cursor implementation for the parser.
+
+use crate::syntax::ast::token::Token;
+
+/// Token cursor.
+///
+/// This internal structure gives basic testable operations to the parser.
+#[derive(Debug, Clone, Default)]
+pub(super) struct Cursor {
+    /// The tokens being input.
+    tokens: Vec<Token>,
+    /// The current position within the tokens.
+    pos: usize,
+}
+
+impl Cursor {
+    /// Creates a new cursor.
+    pub(super) fn new(tokens: Vec<Token>) -> Self {
+        Self {
+            tokens,
+            ..Self::default()
+        }
+    }
+
+    /// Retrieves the current position of the cursor in the token stream.
+    pub(super) fn pos(&self) -> usize {
+        self.pos
+    }
+
+    /// Moves the cursor to the given position.
+    ///
+    /// This is intended to be used *always* with `Cursor::pos()`.
+    ///
+    /// # Example:
+    /// ```
+    /// # let mut cursor = Cursor::new(Vec::new());
+    /// let pos_save = cursor.pos();
+    /// // Do some stuff that might change the cursor position...
+    /// cursor.seek(pos_save);
+    /// ```
+    pub(super) fn seek(&mut self, pos: usize) {
```

Comment:
Maybe it would make sense to abstract over this and provide a stack with breakpoints. So that we just say `cursor.create_breakpoint()` and then `cursor.to_last_breakpoint()`. But we need to remember to go back in all cases, or to remove the breakpoint if we won't use it anymore.

Reply:
So its worth pointing out that in future i would like to have the lexer send the parser tokens through a channel (most likely Crossbeam), this will allow us to have concurrent parsing (or streaming parsing as V8/Spidermonkey call it).

So you need to imaging we may not even have all the tokens yet, in which case `seek` may fail.

I don't think we need to change things around now, but its worth building with that feature in mind.

Reply:
Moved into an issue here:
https://github.com/jasonwilliams/boa/issues/288

Reply:
> Maybe it would make sense to abstract over this and provide a stack with breakpoints. So that we just say `cursor.create_breakpoint()` and then `cursor.to_last_breakpoint()`. But we need to remember to go back in all cases, or to remove the breakpoint if we won't use it anymore.

I think `seek()`  is fine for now